examples/llava/mtmd.h

```diff
@@ -23,60 +29,119 @@
 MTMD_API mtmd_context * mtmd_init_from_file(const char * mmproj_fname,
-                                                const llama_model * text_model,
-                                                const mtmd_context_params ctx_params);
+                                            const struct llama_model * text_model,
+                                            const struct mtmd_context_params ctx_params);
 
 MTMD_API void mtmd_free(mtmd_context * ctx);
 
+// whether we need to set non-causal mask before llama_decode
+MTMD_API bool mtmd_decode_use_non_causal(mtmd_context * ctx);
+
+// whether the current model use M-RoPE for llama_decode
+MTMD_API bool mtmd_decode_use_mrope(mtmd_context * ctx);
+
+
+// mtmd_bitmap
+//
+// length of data must be nx * ny * 3
+// the data is in RGBRGBRGB... format
+MTMD_API mtmd_bitmap *         mtmd_bitmap_init    (uint32_t nx,
+                                                    uint32_t ny,
+                                                    const unsigned char * data);
+MTMD_API uint32_t              mtmd_bitmap_get_nx  (const mtmd_bitmap * bitmap);
+MTMD_API uint32_t              mtmd_bitmap_get_ny  (const mtmd_bitmap * bitmap);
+MTMD_API const unsigned char * mtmd_bitmap_get_data(const mtmd_bitmap * bitmap);
+MTMD_API void                  mtmd_bitmap_free    (mtmd_bitmap * bitmap);
+// bitmap ID is optional, but useful for KV cache tracking
+// these getters/setters are dedicated functions, so you can for example calculate the hash of the image based on mtmd_bitmap_get_data()
+MTMD_API const char * mtmd_bitmap_get_id  (const mtmd_bitmap * bitmap);
+MTMD_API void         mtmd_bitmap_set_id  (mtmd_bitmap * bitmap, const char * id);
+
+
+// mtmd_input_chunks
+//
+// this is simply a list of mtmd_input_chunk
+// the elements can only be populated via mtmd_tokenize()
+MTMD_API mtmd_input_chunks * mtmd_input_chunks_init(void);
+MTMD_API size_t              mtmd_input_chunks_size(mtmd_input_chunks * chunks);
+MTMD_API mtmd_input_chunk *  mtmd_input_chunks_get (mtmd_input_chunks * chunks, size_t idx);
+MTMD_API void                mtmd_input_chunks_free(mtmd_input_chunks * chunks);
+
+// mtmd_input_chunk
+//
+// the instance will be constructed via mtmd_tokenize()
+// it will be freed along with mtmd_input_chunks
+MTMD_API enum mtmd_input_chunk_type mtmd_input_chunk_get_type        (const mtmd_input_chunk * chunk);
+MTMD_API const llama_token *        mtmd_input_chunk_get_tokens_text (const mtmd_input_chunk * chunk, size_t * n_tokens_output);
+MTMD_API const mtmd_image_tokens *  mtmd_input_chunk_get_tokens_image(const mtmd_input_chunk * chunk);
+
+// in case you want to use custom logic to handle the chunk (i.e. KV cache management)
+// you can move the chunk ownership to your own code
+// this will release the chunk from the list of input chunks
+// remember to free the chunk when you are done with it
+MTMD_API mtmd_input_chunk * mtmd_input_chunk_release(mtmd_input_chunk * chunk);
```

Comment:
I am not sure about the usage of this function, but maybe a better name would be `mtmd_input_chunk_take`. But generally it seems like something that should not be needed.

Reply:
Indeed, this API is inspired by `unique_ptr::release()`, which transfer the ownership of a chunk from its container to user code.

For the usage, it's actually related to [this discussion](https://github.com/ggml-org/llama.cpp/pull/12898#discussion_r2071783127) on llama-server. Basically the idea here is to have a mapping `std::map<llama_pos, mtmd_input_chunk *>` to map the **image** chunk to the correct position in `llama_tokens` array. And since I've been played around with the same idea on wllama, I'm pretty sure that this is what we want to have.

Reply:
Since it is making a copy rather than really releasing it, would `mtmd_input_chunk_dup` or `copy` be more accurate?

Reply:
Ignore that, I see that it is moved and only a small struct is allocated. Wouldn't this leave the `mtmd_input_chunks` that used to own this chunk in a bad state?

Reply:
I implemented this in https://github.com/ggml-org/llama.cpp/pull/13184/commits/6bc7a30af5cf87a54b70a730450e95930d3a1adb